### ISSUE TITLE
Active status jumps to last window when it should stay with the current window

### ISSUE THREAD

**Issue by scfarley** [itchyny/lightline.vim#556]
Saving a buffer, that is not the last window, if using `vim-go` and `lightline.vim` causes the status line for the last window to become the active status line.  That is how I first noticed it, but I have two ways to reproduce it.  Typically, I use a plugin manager, but I wanted the test cases to be as minimal as possible.

- Scenario 1
  - `vimrc1`
  ```vim
  set runtimepath^=~/.vim/plugged/lightline.vim
  set runtimepath^=~/.vim/plugged/vim-go
  let g:go_fmt_experimental = 1
  filetype plugin on
  ```
  - Run `vim -u vimrc1 -O file1.go file2.go file3.go`
  - Exec in the first window:  `:w`

- Scenario 2
  - `vimrc2`
  ```vim
  set runtimepath^=~/.vim/plugged/lightline.vim
  filetype plugin on
  ```
  - Run `vim -u vimrc2 -O file1.go file2.go file3.go`
  - Exec in the first window: `doautoall SessionLoadPost`

At least in the first scenario, I think it is tied to the use by `vim-go` of `mkview()/loadview()` when `g:go_fmt_experimental` is enabled.

This may be related to:
- #435
- vim-airline/vim-airline#1444

**Comment by itchyny** [comment]
What is your Vim version? If it is older than [v8.2.0996](https://github.com/vim/vim/commit/40a019f157), can you test with the latest version?

**Comment by scfarley** [comment]
The version would be useful, yes?  :smiley:  It is `v8.2.2569` on FreeBSD.  I also tested `v8.2.2489` on Linux with the same result.

**Comment by itchyny** [comment]
Aha, I remember the problem. This is not the problem of sessions. This is the problem of `doautoall` and we cannot fix by plugin side. Possible solution is to fix Vim to make the event for the current window fire at last.

**Comment by itchyny** [comment]
I just tested vim-airline but it has the same issue. Open multiple windows, switch to the previous window and `doautoall WinEnter` loses the statusline focus. This is an issue of Vim.

**Comment by scfarley** [comment]
The problem I see is `lightline#update()` is being called within the context of each window when the event fires, so `winnr()` will always be the current window in each call.  The last call to `update()` works upon the last window, so it is considered active.  Unfortunately, I do not see a way within that function to ask Vim what was the active window when the event fired.  Is that right?  I speak little Vim script.

This probably breaks other things due to lack of knowledge around the guts of Vim, but I applied the following (and similar to `lightline#enable()`) to get it to "work" for me.

```diff
--- a/plugin/lightline.vim
+++ b/plugin/lightline.vim
@@ -15,7 +15,7 @@ set cpo&vim

 augroup lightline
   autocmd!
-  autocmd WinEnter,BufEnter,BufDelete,SessionLoadPost,FileChangedShellPost * call lightline#update()
+  autocmd WinEnter,BufEnter,BufDelete,FileChangedShellPost * call lightline#update()
   if !has('patch-8.1.1715')
     autocmd FileType qf call lightline#update()
   endif
```

**Comment by itchyny** [comment]
> Unfortunately, I do not see a way within that function to ask Vim what was the active window when the event fired. Is that right?

Vim changes the active window each event invocation in doautoall command, which is the root cause of the issue. This is not limited to `SessionLoadPost` so ignoring this event does not fix the problem.

I sent a patch to Vim (https://github.com/vim/vim/pull/7958) and was included as [8.2.2596](https://github.com/vim/vim/commit/41cd80335cf318c15c8b0139f53ab5e8a02561ef). So in **Vim 8.2.2596** or later, the issue will not happen anymore. There is no workaround for older versions. Please update the version of Vim. The `:doautoall WinEnter` command keeps the active status line as it is.

Ping #448, which is related to this issue.

**Comment by scfarley** [comment]
I realized my "fix" (removing `SessionLoadPost`) broke things, but it stopped the status from switching to the last window, which let me survive.  :grin:

Thank you very much for the real fix to Vim.  I tested against `v8.2.2601` with a clean checkout of `lightline`, and the issue is gone.
